### ISSUE TITLE
A null compound index value breaks Safari >= 10.1 (and iOS 10.3 beta)

### ISSUE THREAD

**Issue by Ramblurr** [dfahlander/Dexie.js#494]
If you specify a compound index, and then subsequently attempt to store a record in IDB and the value for that index is null, Safari >= 10.1 blows up with the error message: 

```UnknownError – "An unknown error occurred within Indexed Database."```

This affects the iOS 10.3 beta as well.

On Chrome and Firefox, there is no error.

##  Code Sample

```js
      db.version(1).stores({
        friends: "++id,name,age,shoeSize,[name+shoeSize]"
      });

      // ...
      // elsewhere in a transaction, this will except
      db.friends.add({name: "Mark", age: 29, shoeSize: null});
```

## Test case

* source: https://github.com/Ramblurr/dexie-null-compound-index/blob/master/test-case.html
* run directly: https://rawgit.com/Ramblurr/dexie-null-compound-index/master/test-case.html

**Comment by dfahlander** [comment]
Thanks for detecting this issue with Safari Beta 3. Should also be filed at https://bugs.webkit.org/

Pinging @beidson.

**Comment by Ramblurr** [comment]
@dfahlander I realize it is probably a browser bug, however I'm not familiar enough with the IDB api to write up a proper bug report. I'm only familiar with the Dexie API ;-)

A quick glance at the [mozilla IDB page](https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API/Using_IndexedDB) doesn't seem to show anything like the compound indices that Dexie supports.

**Comment by beidson** [comment]
> I would be happy to submit a bug report to apple, but I imagine the apple devs want a pure IDB test case.

I'm the Apple dev who works on IDB. I want you to file this at bugs.webkit.org as @dfahlander mentioned. :)

A pure IDB test case would be ideal, but we can work with the Dexie one as well.

Are you positive the exact same test case works in Safari 10.0?
I don't mean "does it not throw an error?" but "are the results what you'd actually expect"?

**Comment by Ramblurr** [comment]
@beidson @dfahlander Done! I've posted the bug over at https://bugs.webkit.org/show_bug.cgi?id=170000

I've updated the test case to demonstrate how the "correct" behavior works as expected.

**Comment by dfahlander** [comment]
Nice bug id you got 🥇

**Comment by Ramblurr** [comment]
Thanks @dfahlander, maybe it will bring some luck ;-)
---
I just found another possibly related bug that is exposed with this test case.

The first time you load the test case it will fail in the affected browsers, then if you reload it, the test hangs. If you open a new tab with the test case, the test case will run.

AFAICT, it seems that once this bug is triggered, after a refresh in the same tab, any calls to IDB (opening a DB, deleting a db) hangs completely.

**Comment by beidson** [comment]
Haven't had a chance to look yet, but the error string mentioned plus the behavior Ramblurr just described sound like a "Databases" process crash.

Check your crash logs and attach to the bugzilla if I'm right.  :)

**Comment by Ramblurr** [comment]
You're right @beidson, there are some crash logs. I've attached them. Thanks for the hint.

**Comment by Ramblurr** [comment]
A patch has been pushed to the webkit repo, thanks @beidson 

Is there a way I can  know when a Safari Tech Preview (on macOS or iOS beta channels) is available containing the fix, so we can test our app?

**Comment by jeromejesse** [comment]
We have a crash too on the Dexie toArray function. If we use this function only for get all entries on a table IndexedDB, the webview crash on iOS 10.3.  

On iOS < 10.3 all is fine.

We suppose the patch fix this.

**Comment by seanbriceland** [comment]
Same issue as @jeromejesse above, working on getting some sort of trace to you although our entire app & debugger crash.

**Comment by beidson** [comment]
> Same issue as @jeromejesse above, working on getting some sort of trace to you although our entire app & debugger crash.

As already mentioned, we already fixed this in WebKit, so don't kill yourself trying to gather more data.

Try your case out in STP 27 or a WebKit nightly to verify.

**Comment by seanbriceland** [comment]
@beidson, first off thanks a bunch for responding so quickly to everything here. How would you recommend running a Cordova built iOS app with a nightly WebKit build or STP 27?

**Comment by ciesielskico** [comment]
I also would like to know about an option to try the nightly webkit with Cordova

**Comment by fix2015** [comment]
is it still actually, because i have bug "An unknown error occurred within Indexed Database." on safari 11, with dexie 2.0

**Comment by dfahlander** [comment]
@fix2015 It's hard to tell if this bug is the reason. Would have to narrow it down. One thing that has been reported errors in Safari lately, has been when a DB was created in previous browser version, then upgraded.

Therefore, I'd recommend you to narrow down the issue and reproduce it. If it's not reproducable when re-creating the DB, it might be a migration bug as mentioned above. If so, those can also be reproduced even though it's harder to test (for example 1) install Safari version X, run this HTML page, 2) upgrade Safari to version Y).

**Comment by dfahlander** [comment]
Heads-up: Safari seems to have another binary-keys related bug that makes IDB unresponsive until restarting Safari. @fix2015 has filed the bug separately in #656.
